File: cortex/tests/fixtures/ros2_workspace/src/demo_pkg/src/listener.cpp
// Stub ROS2 C++ source used only by kg tests; not built.
//
// Exercises the subscription / lifecycle-node / client subset of the
// ``ros2_topology`` (C++ half) extractor:
//   - rclcpp_lifecycle::LifecycleNode subclass (lifecycle: true prop)
//   - create_subscription<sensor_msgs::msg::Image>("camera/image", ...)
//   - rclcpp_action::create_client<demo_pkg::action::Fibonacci>(...)
//   - declare_parameter<std::string>("frame_id", "camera_link")

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_lifecycle/lifecycle_node.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "demo_pkg/action/fibonacci.hpp"

namespace demo_pkg {

class Listener : public rclcpp_lifecycle::LifecycleNode {
 public:
  Listener() : LifecycleNode("listener") {
    subscription_ = create_subscription<sensor_msgs::msg::Image>(
        "camera/image", 10, [](sensor_msgs::msg::Image::SharedPtr) {});
    action_client_ = rclcpp_action::create_client<demo_pkg::action::Fibonacci>(
        this, "fibonacci");
    declare_parameter<std::string>("frame_id", "camera_link");
  }

 private:
  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr subscription_;
  rclcpp_action::Client<demo_pkg::action::Fibonacci>::SharedPtr action_client_;
};

}  // namespace demo_pkg

File: weld/tests/fixtures/ros2_workspace/src/demo_pkg/demo_pkg/listener.py
# Stub rclpy source used only by Weld tests; not executed.
#
# Exercises the subscription / lifecycle-node / action-client subset of
# the ``ros2_topology`` (Python half) extractor:
#
# - ``rclpy.lifecycle.LifecycleNode`` subclass (lifecycle: True prop)
# - ``self.create_subscription(sensor_msgs.msg.Image, "camera/image", ...)``
# - ``ActionClient(self, demo_pkg.action.Fibonacci, "fibonacci")``
# - ``self.declare_parameter("frame_id", "camera_link")``

from rclpy.action import ActionClient
from rclpy.lifecycle import LifecycleNode

import demo_pkg.action
import sensor_msgs.msg

class Listener(LifecycleNode):
    def __init__(self) -> None:
        super().__init__("listener")
        self.subscription_ = self.create_subscription(
            sensor_msgs.msg.Image,
            "camera/image",
            self._on_image,
            10,
        )
        self.action_client_ = ActionClient(
            self, demo_pkg.action.Fibonacci, "fibonacci"
        )
        self.declare_parameter("frame_id", "camera_link")

    def _on_image(self, msg) -> None:  # pragma: no cover - fixture only
        del msg
